docs/notes/threading-info.txt

```diff
@@ -0,0 +1,563 @@
+qdr_core_t's action_list).  The core thread blocks on a mutex
+condition waiting for work.  When a worker thread needs a service from
+the core thread, the worker queues a work item to the action_list then
+triggers the condition.  This causes the core thread to unblock and
+service the action_list work items.
+
+
+Source Code Conventions
+=======================
+
+In general, core thread code and data structure have names that are
+prefixed by "qdr_".  Functions that can only be run on the core thread
+- that cannot be invoked from the context of a worker thread - are
+suffixed with "_CT".  For example "qdr_update_delivery_CT()" is a core
+function that must never be invoked by a worker thread.
+
+Code that is private to the Core thread is in the "src/router_core"
+subdirectory of the source code repo.
+
+Worker thread code does not have a particular naming convention. Most
+worker thread-only APIs do start with "qd_", but this prefix is not
+reserved for worker thread APIs.
+
+
+Principal Data Structures
+=========================
+
+There are many data structures that are used throughout the code.
+This section will overview a few of those data structures which are
+involved in inter-thread communications.
+
+qdr_delivery_t
+--------------
+
+The qdr_delivery_t structure represents the router's view of a Proton
+delivery instance (pn_delivery_t).  Its definition is private to core,
+however core defines an API that allows a worker to concurrently
+access parts of this structure.
+
+There is a one-to-one relationship between a qdr_delivery_t and a
+Proton pn_delivery_t.  When an incoming message arrives at the router
+a new incoming qdr_delivery_t is created and tied to the message's
+pn_delivery_t.  When a message is forwarded a new outgoing
+qdr_delivery_t is created to associate with the outgoing link's
+pn_delivery_t.
+
+It is important to understand that pn_delivery_t's - and by extension
+qdr_delivery_t's - are "owned" by the parent pn_link_t
+(qdr_link_t). This means that a delivery object cannot be moved from
+an inbound link to an outbound link - which seems counterintiutive to
+the process of routing a message.
+
+In reality when an inbound delivery is routed, the router create a new
+outbound delivery for each outbound link.  These outbound deliveries
+are also represented by their own pn_delivery_t and qdr_delivery_t
+instances.
+
+The core thread links the inbound and outbound qdr_deliver_t's
+together.  This makes it trival to determine the "peer" of any
+delivery that is being forwarded.
+
+Note that a single inbound qdr_delivery_t may be linked to
+more than one outgoing qdr_delivery_t in the case of multicast
+forwarding.
+
+A qdr_delivery_t state includes:
+
+ *) the delivery's disposition (local and remote)
+ *) the delivery's settlement state
+ *) a reference to the AMQP message (qd_message_t *)
+
+The qdr_delivery_t structure is defined in src/router_core/delivery.h
+
+qd_message_t
+------------
+
```

Comment:
You should mention that qdrouterd does not use pn_message_t and has created qd_message_t as a replacement.  This is in contrast to qdr_delivery_t being associated with pn_delivery_t.